Release/src/http/client/http_client_asio.cpp

```diff
@@ -1966,6 +1966,124 @@ void asio_client::send_request(const std::shared_ptr<request_context>& request_c
+        // automatic redirection to that URI."
+        return true;
+    case status_codes::PermanentRedirect:
+        // This status code "does not allow changing the request method from POST
+        // to GET."
+        return true;
+    default:
+        return false;
+    }
+}
+
+static bool is_recognized_redirection(status_code code)
+{
+    // other 3xx status codes are not handled
+    return is_retrieval_redirection(code) || is_unchanged_redirection(code);
+}
+
+static bool is_retrieval_request(method method)
+{
+    return methods::GET == method || methods::HEAD == method;
+}
+
+static const std::vector<utility::string_t> request_body_header_names =
+{
+    header_names::content_encoding,
+    header_names::content_language,
+    header_names::content_length,
+    header_names::content_location,
+    header_names::content_type
+};
+
+pplx::task<http_response> follow_redirect(http_client_config config,
+                                          http_request request,
+                                          http_response response,
+                                          std::vector<uri> followed_urls = {})
+{
+    // Return immediately if the response is not a supported redirection
+    if (!is_recognized_redirection(response.status_code()))
+        return pplx::task_from_result(response);
+
+    // Although not required by RFC 7231, config may limit the number of automatic redirects
+    if (config.max_redirects() <= followed_urls.size())
+        return pplx::task_from_result(response);
+
+    // Can't very well automatically redirect if the server hasn't provided a Location
+    if (!response.headers().has(header_names::location))
+        return pplx::task_from_result(response);
+
+    const uri location(request.absolute_uri().resolve_uri(response.headers()[header_names::location]));
+
+    // Config may prohibit automatic redirects from HTTPS to HTTP
+    if (!config.https_to_http_redirects() && request.absolute_uri().scheme() == _XPLATSTR("https")
+        && location.scheme() != _XPLATSTR("https"))
+        return pplx::task_from_result(response);
+
+    // "A client SHOULD detect and intervene in cyclical redirections."
+    if (followed_urls.end() != std::find(followed_urls.begin(), followed_urls.end(), location))
+        return pplx::task_from_result(response);
+
+    // Implementation limitation: cannot redirect e.g. a POST request using the same method
+    // since the request body may have been consumed
+    if (!is_retrieval_request(request.method()) && !is_retrieval_redirection(response.status_code()))
+        return pplx::task_from_result(response);
+
+    const auto method = is_retrieval_request(request.method()) ? request.method() : methods::GET;
+
+    http_request redirect(method);
+
+    // Generally, sending the same headers as the original request makes sense
+    // However, Content-* headers do not make sense if no body is being sent
+    redirect.headers() = request.headers();
```

Comment:
Instead of copying all the headers and then removing selectively, does it make sense to loop through request.headers() and move the content out into the redirect's headers block if the header isn't in request_body_header_names?

Reply:
My concern with that was that the original request object can 'belong' to user code, so destroying the headers in that seemed wrong?

Reply:
Aren't you eating the request here though? I don't see where it's exposed to user code but I could have just missed something. (As I said this is a super nitpick though)

Reply:
Oh derp, *request*, I was thinking response. It's passed by value into this function though right? So I think we already have our own copy...

Reply:
It's only a shallow copy though, `shared_ptr` impl?

Reply:
Ah, you're right. :(

You never look as stupid as when making offhand comments because you've been put in the maintainer shoes for a product you don't really understand.... :)

Reply:
:)

Even looping through `request.headers()` and copying the content into the redirect's headers isn't optimal because `http_headers` doesn't have a simple `insert` only a complicated `add`... Sigh.